homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,1133 @@
+        disk.  It also does some basic checking to make sure relative paths
+        outside of the legitimate locations cn't be used """
+        entity_id = path_match.group('entity_id')
+
+        camera = None
+        if entity_id in component.entities.keys():
+            camera = component.entities[entity_id]
+
+        if camera:
+            image_path = os.path.normpath(os.path.join(
+                camera.images_path, data['image_path']))
+
+            # Check to see that someone is not trying to do something dodgey
+            # with relative paths
+            if not image_path.startswith(camera.images_path):
+                handler.send_response(HTTP_NOT_FOUND)
+
+            handler.write_file(image_path)
+
+        else:
+            handler.send_response(HTTP_NOT_FOUND)
+
+    hass.http.register_path(
+        'GET',
+        re.compile(r'/api/saved_camera_image/(?P<entity_id>[a-zA-Z\._0-9]+)'),
+        _saved_camera_image,
+        require_auth=True)
+
+    def handle_motion_detection_service(service):
+        """ Handles calls to the camera services. """
+
+        action = service.data.get('action', None)
+        if action is None:
+            logging.getLogger(__name__).warning(
+                "Call to camera service did not specify an \
+                 action in the service data")
+            return
+
+        state = service.data.get('state', None)
+        if state is None:
+            logging.getLogger(__name__).warning(
+                "Call to camera service did not specify a \
+                 state in the service data")
+            return
+
+        target_cameras = component.extract_from_service(service)
+
+        for camera in target_cameras:
+            if action == SWITCH_ACTION_MOTION:
+                if (state == STATE_ON and not
+                        camera.is_motion_detection_enabled):
+                    camera.enable_motion_detection()
+                elif (
+                        state == STATE_OFF and
+                        camera.is_motion_detection_enabled):
+                    camera.disable_motion_detection()
+
+            elif action == SWITCH_ACTION_RECORD:
+                if (state == STATE_ON and not
+                        camera.is_recording):
+                    camera.record_stream()
+                elif (
+                        state == STATE_OFF and
+                        camera.is_recording):
+                    camera.stop_recording()
+
+            elif action == SWITCH_ACTION_SNAPSHOT:
+                if (state == STATE_ON and not
+                        camera.is_taking_snapshot):
+                    camera.take_snapshot()
+
+            camera.update_ha_state(True)
+
+    hass.services.register(
+        DOMAIN,
+        SERVICE_CAMERA,
+        handle_motion_detection_service)
+
+    return True
+
+
+# pylint: disable=too-many-instance-attributes
+# pylint: disable=too-many-public-methods
+class Camera(Entity):
```

Comment:
The comments for the methods should specify which ones we expect to be overwritten by platforms and which not.